### PR TITLE
Add support for Tomcat 8.5

### PR DESCRIPTION
**Pull request by martin-flower**
I have attempted to add support for Tomcat 8.5 - it is likely that this work will need some finishing off.  See comments in README.md and RELEASE_NOTES.md.

### CODE REVIEW

**Comment by bmuschko** [comment]
Thanks for providing the pull request. I am going to review soon.

File: README.md [hunk 1]
```diff
@@ -128,21 +128,37 @@ dependencies {
...
+
+```groovy
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    def tomcatVersion = '8.5.12'
+    tomcat "org.apache.tomcat.embed:tomcat-embed-core:${tomcatVersion}",
+           "org.apache.tomcat.embed:tomcat-embed-logging-juli:8.5.2",
```
Comment (bmuschko) [inline README.md]: Is this dependency even required with Tomcat 8.5. I couldn't find anything on Apache's documentation page? If it is required why does it have a different version than the rest of the dependencies?

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -1,3 +1,12 @@
+### Version 2.2.6 (2017)
+
+* Add Tomcat 8.5.x alongside 8.0.x
+* Updated tomcat versions to latest release version
+* Requires tomcat configuration in build.gradle:
```
Comment (bmuschko) [inline RELEASE_NOTES.md]: Why exactly is that the case?

File: embedded/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/TomcatVersion.groovy [hunk 3]
```diff
@@ -21,9 +21,14 @@ package com.bmuschko.gradle.tomcat.embedded
...
-    VERSION_7X('7.0', 'Tomcat 7x'),
-    VERSION_8X('8.0', 'Tomcat 8x')
+
+    // @see https://wiki.apache.org/tomcat/TomcatVersions
+
+    VERSION_6_0_X('6.0' , 'Tomcat 6.0.x'),
+    VERSION_7_0_X('7.0' , 'Tomcat 7.0.x'),
+    VERSION_8_0_X('8.0' , 'Tomcat 8.0.x'),
+    VERSION_8_5_X('8.5',  'Tomcat 8.5.x')
+//  VERSION_9_0_X('9.0' , 'Tomcat 9.0.x')
```
Comment (bmuschko) [inline embedded/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/TomcatVersion.groovy]: Let's leave out version 9.x for now. We can make that part of another PR.

File: plugin/src/functTest/groovy/com/bmuschko/gradle/tomcat/TomcatJasperFunctionalTest.groovy [hunk 4]
```diff
@@ -30,7 +30,7 @@ class TomcatJasperFunctionalTest extends AbstractFunctionalTest {
         new File(compiledJspDir, 'date_jsp.java').exists()
 
         where:
-        tomcatVersion << [TomcatVersion.VERSION_6X, TomcatVersion.VERSION_7X, TomcatVersion.VERSION_8X]
+        tomcatVersion << [TomcatVersion.VERSION_6_0_X, TomcatVersion.VERSION_7_0_X, TomcatVersion.VERSION_8_0_X] // TODO , TomcatVersion.VERSION_8_5_X
```
Comment (bmuschko) [inline plugin/src/functTest/groovy/com/bmuschko/gradle/tomcat/TomcatJasperFunctionalTest.groovy]: Did that not work or why is it commented out for 8.5.x?

File: tomcat85x/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/Tomcat85xServer.groovy [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import java.lang.reflect.Constructor
+
+/**
+ * Tomcat 85x server implementation.
+ *
+ * @author Benjamin Muschko
+ * @author Andrey Bloschetsov
+ * @author Martin Flower
+ */
+class Tomcat85xServer extends BaseTomcat7xPlusImpl {
```
Comment (bmuschko) [inline tomcat85x/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/Tomcat85xServer.groovy]: I don't see any difference to `Tomcat8xServer` except the version. Can't we just extend from that class and only override `getVersion()`?

File: RELEASE_NOTES.md (hunk 2 above)
Comment (bmuschko) [inline RELEASE_NOTES.md]: If it is something specific to Tomcat 8.5.x why not make that the default in the Tomcat 8.5.x implementation?

File: build.gradle [hunk 6]
```diff
@@ -17,7 +17,7 @@ allprojects {
     apply plugin: 'groovy'
     apply plugin: 'idea'
     group = 'com.bmuschko'
-    version = '2.2.5'
+    version = '2.2.6'
```
Comment (bmuschko) [inline build.gradle]: I think the functionality calls for a minor version upgrade rather than a patch upgrade.

File: README.md (hunk 1 above)
Comment (martin-flower) [inline README.md]: Hi @bmuschko, tomcat-embed-logging-juli - the latest version on Maven Central is 8.5.2. I don't know if it is required - I have just updated the existing documentation.

File: RELEASE_NOTES.md (hunk 2 above)
Comment (martin-flower) [inline RELEASE_NOTES.md]: Are you referring to httpProtocol and ajpProtocol? These are the parameters that worked for me. It would seem natural to make them the defaults. I don't know Tomcat or the plugin well enough in order to comment on this.

File: embedded/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/TomcatVersion.groovy (hunk 3 above)
Comment (martin-flower) [inline embedded/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/TomcatVersion.groovy]: yes

File: plugin/src/functTest/groovy/com/bmuschko/gradle/tomcat/TomcatJasperFunctionalTest.groovy (hunk 4 above)
Comment (martin-flower) [inline plugin/src/functTest/groovy/com/bmuschko/gradle/tomcat/TomcatJasperFunctionalTest.groovy]: TODO , TomcatVersion.VERSION_8_5_X - test failed

File: tomcat85x/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/Tomcat85xServer.groovy (hunk 5 above)
Comment (martin-flower) [inline tomcat85x/src/main/groovy/com/bmuschko/gradle/tomcat/embedded/Tomcat85xServer.groovy]: Hmm, from memory I think that is where I started, but I ran into problems later on. These were probably related to me not understanding the plugin. My expectation is that everything should work by just extending BaseTomcat7xPlusImpl.

File: build.gradle (hunk 6 above)
Comment (martin-flower) [inline build.gradle]: 2.3.0 instead of 2.2.6

**Comment by martin-flower** [comment]
> Did you build a new version of the plugin locally and try it out with a web project for Tomcat 8.5.x?

Yes, I built a local 2.2.6 and tested it locally.

**Comment by martin-flower** [comment]
> Can you please added the comments I added?

By 'comments' do you mean your comments on this page, or is there another pull request I have overlooked?

**Comment by wodencafe** [comment]
Hi @bmuschko and @martin-flower , what needs to be done to get this PR merged?

This functionality would really help us in our deployment testing.

**Comment by bmuschko** [comment]
@wodencafe There are still some open questions from my end as indicated by my comments. We should find out _why_ a certain configuration works before just setting it or proposing it to users e.g. https://github.com/bmuschko/gradle-tomcat-plugin/pull/156#discussion_r111158935 or https://github.com/bmuschko/gradle-tomcat-plugin/pull/156#discussion_r111159605. If tests fail then we'll need to figure out why they fail. It's likely that users will run into the same issues.

I just haven't gotten around to looking into them myself. @wodencafe Feel free to investigate further. I hoping to have a look within the next 2-3 weeks.

**Comment by bmuschko** [comment]
Thanks for providing the pull request. It has been merged. I addressed my comments after merging.
